tests/nested/manual/refresh-revert-fundamentals/task.yaml

```diff
@@ -0,0 +1,137 @@
+    . "$TESTSLIB/nested.sh"
+
+    if [ -f skip.test ]; then
+        rm -f skip.test
+        exit
+    fi
+
+    destroy_nested_vm
+    cleanup_nested_env
+
+    rm -f "$WORK_DIR/image/ubuntu-core.img"
+
+execute: |
+    #shellcheck source=tests/lib/nested.sh
+    . "$TESTSLIB/nested.sh"
+
+    if [ -f skip.test ]; then
+        exit
+    fi
+
+    FROM_REV="$(get_snap_rev_for_channel "$SNAP" $TRACK/$CORE_CHANNEL)"
+    TO_REV="$(get_snap_rev_for_channel "$SNAP" $TRACK/$CORE_REFRESH_CHANNEL)"
+
+    execute_remote "snap list $SNAP" | MATCH "^${SNAP}.*${FROM_REV}.*${TRACK}/${CORE_CHANNEL}.*"
+    # The snap is refreshed
+    REFRESH_ID=$(execute_remote "sudo snap refresh --no-wait --channel $CORE_REFRESH_CHANNEL $SNAP")
+
+    case "$SNAP" in
+        snapd|pc)
+            # we manually reboot even after snapd refresh to ensure that if
+            # resealing took place we are still able to boot
+            execute_remote "snap watch $REFRESH_ID"
+            execute_remote "snap changes" | MATCH "$REFRESH_ID\s+Done\s+.*"
+            execute_remote "sudo reboot"
+            ;;
+        pc-kernel|core20)
+            # don't manually reboot, wait for automatic snapd reboot
+            ;;
+    esac
+    wait_for_no_ssh
+    wait_for_ssh
+
+    # Check the new version of the snaps is correct after the system reboot
+    execute_remote "snap list $SNAP" | MATCH "^${SNAP}.*${TO_REV}.*${TRACK}/${CORE_REFRESH_CHANNEL}.*"
+
+    # We check the change is completed
+    case "$SNAP" in
+        pc-kernel|core20)
+            for _ in $(seq 10); do
+                if execute_remote "snap changes" | MATCH "$REFRESH_ID\s+Done\s+.*"; then
+                    break
+                fi
+                sleep 1
+            done
+            execute_remote "snap changes" | MATCH "$REFRESH_ID\s+Done\s+.*"
+            ;;
+    esac
+
+    # The snap is reverted
+    REVERT_ID=$(execute_remote "sudo snap revert --no-wait $SNAP")
+
+    case "$SNAP" in
+        snapd|pc)
+            # we manually reboot even after snapd refresh to ensure that if
+            # resealing took place we are still able to boot
+            execute_remote "snap watch $REVERT_ID"
+            execute_remote "snap changes" | MATCH "$REVERT_ID\s+Done\s+.*"
+            execute_remote "sudo reboot"
+            ;;
+        pc-kernel|core20)
+            # don't manually reboot, wait for automatic snapd reboot
+            ;;
+    esac
+    wait_for_no_ssh
+    wait_for_ssh
+
+    # Check the version of the snaps after the revert is correct
+    execute_remote "snap list $SNAP" | MATCH "^${SNAP}.*${FROM_REV}.*${TRACK}/${CORE_REFRESH_CHANNEL}.*"
+
+    # We check the change is completed
+    case "$SNAP" in
+        pc-kernel|core20)
+            for _ in $(seq 10); do
+                if execute_remote "snap changes" | MATCH "$REVERT_ID\s+Done\s+.*"; then
+                    break
+                fi
+                sleep 1
+            done
```

Comment:
same comment, `retry` seems appropriate here

Reply:
Initially I tried that but if I use retry I can't use execute_remote as it is defined in the nested.sh helper, any idea how to make retry work with a function defined in a helper?

Reply:
ah interesting good point I hadn't thought about that... this is fine for now then, but let's have a chat with @zyga when he's back to see what we can do to make this kind of thing easier, perhaps `retry` can gain an `--includes` like

```bash
retry --includes="$TESTSLIB/nested.sh" -- execute_remote "snap changes" | MATCH "$REFRESH_ID\s+Done\s+.*"
```